lib/ansible/modules/cloud/amazon/s3_bucket.py

```diff
@@ -156,172 +130,185 @@ def _create_or_update_bucket(connection, module, location):
+
     if policy is not None:
         if isinstance(policy, string_types):
             policy = json.loads(policy)
 
-        if not policy:
-            bucket.delete_policy()
-            # only show changed if there was already a policy
-            changed = bool(current_policy)
+        if not policy and current_policy:
+            try:
+                s3_client.delete_bucket_policy(Bucket=name)
+            except (BotoCoreError, ClientError) as e:
+                module.fail_json_aws(e, msg="Failed to delete bucket policy")
+            changed = True
 
         elif compare_policies(current_policy, policy):
-            changed = True
             try:
-                bucket.set_policy(json.dumps(policy))
-                current_policy = json.loads(to_native(bucket.get_policy()))
-            except S3ResponseError as e:
-                module.fail_json(msg=e.message)
+                s3_client.put_bucket_policy(Bucket=name, Policy=json.dumps(policy))
+            except (BotoCoreError, ClientError) as e:
+                module.fail_json_aws(e, msg="Failed to update bucket policy")
+            changed = True
+            current_policy = policy
 
     # Tags
     try:
-        current_tags = bucket.get_tags()
-    except S3ResponseError as e:
-        if e.error_code == "NoSuchTagSet":
+        current_tags = s3_client.get_bucket_tagging(Bucket=name).get('TagSet')
+    except ClientError as e:
+        if e.response['Error']['Code'] == 'NoSuchTagSet':
             current_tags = None
         else:
-            module.fail_json(msg=e.message)
+            module.fail_json_aws(e, msg="Failed to get bucket tags")
+    except BotoCoreError as e:
+        module.fail_json_aws(e, msg="Failed to get bucket tags")
 
     if current_tags is None:
         current_tags_dict = {}
     else:
-        current_tags_dict = dict((t.key, t.value) for t in current_tags[0])
+        current_tags_dict = boto3_tag_list_to_ansible_dict(current_tags)
 
     if tags is not None:
         if current_tags_dict != tags:
-            try:
-                if tags:
-                    bucket.set_tags(create_tags_container(tags))
-                else:
-                    bucket.delete_tags()
-                current_tags_dict = tags
-                changed = True
-            except S3ResponseError as e:
-                module.fail_json(msg=e.message)
+            if tags:
+                try:
+                    s3_client.put_bucket_tagging(Bucket=name, Tagging={'TagSet': ansible_dict_to_boto3_tag_list(tags)})
+                except (BotoCoreError, ClientError) as e:
+                    module.fail_json_aws(e, msg="Failed to update bucket tags")
+            else:
+                try:
+                    s3_client.delete_bucket_tagging(Bucket=name)
+                except (BotoCoreError, ClientError) as e:
+                    module.fail_json_aws(e, msg="Failed to delete bucket tags")
+            current_tags_dict = tags
```

Comment:
Here's a sporadic test failure that a custom waiter would probably fix:
```
TASK [s3_bucket : Do not specify any tag to ensure previous tags are not removed] **********************************************
task path: /Users/shertel/Workspace/ansible/test/integration/targets/s3_bucket/tasks/main.yml:184
File lookup using /Users/shertel/Workspace/ansible/test/integration/targets/s3_bucket/templates/policy.json as file
Using module file /Users/shertel/Workspace/ansible/lib/ansible/modules/cloud/amazon/s3_bucket.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: shertel
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/shertel/.ansible/tmp/ansible-tmp-1521813741.78-18776467509358 `" && echo ansible-tmp-1521813741.78-18776467509358="` echo /Users/shertel/.ansible/tmp/ansible-tmp-1521813741.78-18776467509358 `" ) && sleep 0'
<127.0.0.1> PUT /Users/shertel/.ansible/tmp/ansible-local-98052I5QPSp/tmpmypRU6 TO /Users/shertel/.ansible/tmp/ansible-tmp-1521813741.78-18776467509358/s3_bucket.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/shertel/.ansible/tmp/ansible-tmp-1521813741.78-18776467509358/ /Users/shertel/.ansible/tmp/ansible-tmp-1521813741.78-18776467509358/s3_bucket.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/tmp/ansible-test-coverage-2Ex1l6/coverage/injector.py /Users/shertel/.ansible/tmp/ansible-tmp-1521813741.78-18776467509358/s3_bucket.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /Users/shertel/.ansible/tmp/ansible-tmp-1521813741.78-18776467509358/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "aws_access_key": "redacted",
            "aws_secret_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ceph": false,
            "ec2_url": null,
            "force": false,
            "name": "ansible-test-shertel-osx-46829947-testbucket-ansible-complex",
            "policy": "{\"Version\": \"2012-10-17\", \"Statement\": [{\"Action\": [\"s3:GetObject\"], \"Sid\": \"AddPerm\", \"Resource\": [\"arn:aws:s3:::ansible-test-shertel-osx-46829947-testbucket-ansible-complex/*\"], \"Effect\": \"Allow\", \"Principal\": \"*\"}]}",
            "profile": null,
            "region": "us-east-1",
            "requester_pays": false,
            "s3_url": null,
            "security_token": "",
            "state": "present",
            "tags": null,
            "validate_certs": true,
            "versioning": false
        }
    },
    "name": "ansible-test-shertel-osx-46829947-testbucket-ansible-complex",
    "policy": {
        "Statement": [
            {
                "Action": "s3:GetObject",
                "Effect": "Allow",
                "Principal": "*",
                "Resource": "arn:aws:s3:::ansible-test-shertel-osx-46829947-testbucket-ansible-complex/*",
                "Sid": "AddPerm"
            }
        ],
        "Version": "2012-10-17"
    },
    "requester_pays": false,
    "tags": {
        "example": "tag1-udpated"
    },
    "versioning": {
        "MfaDelete": "Disabled",
        "Versioning": "Suspended"
    }
}

TASK [s3_bucket : assert] ******************************************************************************************************
task path: /Users/shertel/Workspace/ansible/test/integration/targets/s3_bucket/tasks/main.yml:194
fatal: [localhost]: FAILED! => {
    "assertion": "not output.changed",
    "changed": false,
    "evaluated_to": false
}
```
It successfully removed a tag in the previous task, so it should wait for the bucket to reflect the correct tags so the next task is idempotent.